### PR TITLE
BLD: streamlined library names in site.cfg sections

### PR DESCRIPTION
**Pull request by zerothi**
This unifies the names for retrieving libraries in sections.
Previously one could only get umfpack libraries using:

    [umfpack]
    umfpack_libs = ...

while now it prefers to use

    [umfpack]
    libraries = ...

(but defaults to use umfpack_libs).
Some libraries is system_info uses libraries whereas others were using
`*_libs` as keys. All future developments I think should retain the
libraries keyword, and only if applicable for sub-packages it should
be deferred to more `_libs` entries.

This commit also fixes CBLAS linkage if the user has supplied a
sufficient library. I.e. it first checks whether the CBLAS interface
is present *as is*, then tries to add cblas, etc.
It should at least now work for more diverse libraries.

### CODE REVIEW

File: numpy/distutils/system_info.py [hunk 1]
```diff
@@ -1131,7 +1133,9 @@ def calc_info(self):
         info = {}
         atlas_libs = self.get_libs('atlas_libs',
                                    self._lib_names + self._lib_atlas)
+        atlas_libs = self.get_libs('libraries', atlas_libs)
         lapack_libs = self.get_libs('lapack_libs', self._lib_lapack)
+        lapack_libs = self.get_libs('lapack_libraries', lapack_libs)
```
Comment (eric-wieser) [inline numpy/distutils/system_info.py]: Why a third spelling?

File: numpy/distutils/system_info.py [hunk 2]
```diff
@@ -1716,34 +1725,43 @@ def has_cblas(self, info):
...
                 obj = c.compile([src], output_dir=tmpdir,
                                 include_dirs=self.get_include_dirs())
-
-                # check we can link (find library)
-                # some systems have separate cblas and blas libs. First
-                # check for cblas lib, and if not present check for blas lib.
+            except (distutils.ccompiler.CompileError, distutils.ccompiler.LinkError):
+                res = None
+                # skip to finally clause, we can't compile anyway!
+                raise PassToFinal
```
Comment (eric-wieser) [inline numpy/distutils/system_info.py]: Just use `return None` here

File: numpy/distutils/system_info.py (hunk 1 above)
Comment (zerothi) [reply numpy/distutils/system_info.py]: Agreed!

File: numpy/distutils/system_info.py (hunk 2 above)
Comment (zerothi) [reply numpy/distutils/system_info.py]: Indeed that is cleaner!

File: numpy/distutils/system_info.py [hunk 3]
```diff
@@ -1724,26 +1732,26 @@ def has_cblas(self, info):
...
                 try:
                     c.link_executable(obj, os.path.join(tmpdir, "a.out"),
-                                      libraries=["cblas"],
+                                      libraries=libs,
                                       library_dirs=info['library_dirs'],
                                       extra_postargs=info.get('extra_link_args', []))
-                    res = "cblas"
+                    res = libs
+                    # This breaks the for loop
+                    break
```
Comment (eric-wieser) [inline numpy/distutils/system_info.py]: Just use `return libs` here

File: numpy/distutils/system_info.py [hunk 4]
```diff
@@ -1683,6 +1689,7 @@ class blas_info(system_info):
     def calc_info(self):
         lib_dirs = self.get_lib_dirs()
         blas_libs = self.get_libs('blas_libs', self._lib_names)
+        blas_libs = self.get_libs('libraries', blas_libs)
```
Comment (eric-wieser) [inline numpy/distutils/system_info.py]: Does this merge the results or discard one half off them if both are specified?

File: numpy/distutils/system_info.py [hunk 5]
```diff
@@ -1698,14 +1705,15 @@ def calc_info(self):
             lib = self.has_cblas(info)
             if lib is not None:
                 info['language'] = 'c'
-                info['libraries'] = [lib]
+                info['libraries'] = lib
                 info['define_macros'] = [('HAVE_CBLAS', None)]
         self.set_info(**info)
 
     def has_cblas(self, info):
+        """ changed in 1.17 to return a list instead of a str """
```
Comment (eric-wieser) [inline numpy/distutils/system_info.py]: Can you move this change to another PR? The changes to this function seem unrelated to the renaming you do above.

File: numpy/distutils/system_info.py (hunk 4 above)
Comment (zerothi) [reply numpy/distutils/system_info.py]: It has `libraries` precedence. So if only one is present, only that one will be used. I.e.:

```bash
[blas]
blas_libs = foo,bar

[blas]
libraries = foo,bar

[blas]
blas_libs = discarded
libraries = foo,bar
```
will all yield the same.

File: numpy/distutils/system_info.py (hunk 3 above)
Comment (zerothi) [reply numpy/distutils/system_info.py]: done, this implies that the last `return res` also changes to `return None`, for clarity!

File: numpy/distutils/system_info.py (hunk 5 above)
Comment (zerothi) [reply numpy/distutils/system_info.py]: Ok, I'll do that, I'll need to extract the entire CBLAS change then.

Comment (zerothi) [reply numpy/distutils/system_info.py]: See #13161

File: numpy/distutils/system_info.py (hunk 4 above)
Comment (eric-wieser) [reply numpy/distutils/system_info.py]: In that case, you should emit an error if both are specified.

Comment (zerothi) [reply numpy/distutils/system_info.py]: If both are specified and we should raise an error it would require some additional restructuring which I don't think is necessary.
I would be against this since many of the `site.cfg` are undocumented features anyways. I.e. people would probably in some overlapping period expect the same `site.cfg` to work for both cases and when the older `*_libs` finally gets deprecated *then* one could do something.

Do you still think both should raise an error?

Comment (zerothi) [reply numpy/distutils/system_info.py]: Additionally I think this is of very minor relevance to 99.9% of users since the `site.cfg` is generally generated with one configuration in mind.

Comment (eric-wieser) [reply numpy/distutils/system_info.py]: >  are undocumented features anyways

All the more reason to raise a clear error if they do something that's a bad idea

> Do you still think both should raise an error?

Yes, silently discarding information should definitely raise an error.

File: numpy/distutils/system_info.py [hunk 6]
```diff
@@ -1131,6 +1133,7 @@ def calc_info(self):
         info = {}
         atlas_libs = self.get_libs('atlas_libs',
                                    self._lib_names + self._lib_atlas)
+        atlas_libs = self.get_libs('libraries', atlas_libs)
         lapack_libs = self.get_libs('lapack_libs', self._lib_lapack)
```
Comment (eric-wieser) [inline numpy/distutils/system_info.py]: Why is `lapack_libs` still used as the `x_libs` format, but `atlas_libs` aliased to `libraries`?

File: numpy/distutils/system_info.py (hunk 4 above)
Comment (zerothi) [reply numpy/distutils/system_info.py]: Ok, I'll fix this

File: numpy/distutils/system_info.py [hunk 7]
```diff
@@ -1877,10 +1884,8 @@ class blis_info(blas_info):
 
     def calc_info(self):
         lib_dirs = self.get_lib_dirs()
-        blis_libs = self.get_libs('libraries', self._lib_names)
-        if blis_libs == self._lib_names:
-            blis_libs = self.get_libs('blis_libs', self._lib_names)
```
Comment (eric-wieser) [inline numpy/distutils/system_info.py]: I hadn't realized we already accepted `libraries` elsewhere. I think using this approach with the `if` statement everywhere makes sense to me.

File: numpy/distutils/system_info.py (hunk 6 above)
Comment (zerothi) [reply numpy/distutils/system_info.py]: Yes, this is a bit problematic with ATLAS. ATLAS can provide either BLAS or BLAS+LAPACK(sub) or BLAS+LAPACK, with the middle option being only a subset of routines which are optimized.
I would suggest *any* user to put all libraries (both BLAS and LAPACK) in a single library entry, however it seems that the dev of the `atlas_info` class recommended LAPACK and BLAS libraries to be split. The only case where this would be troublesome is to use ATLAS BLAS but another LAPACK lib.

Basically I think this is why there is an abstraction of another key for only the LAPACK libraries.

So the final atlas section would look like:
```bash
[atlas]
libraries = f77blas,cblas,atlas
lapack_libs = lapack_atlas
```
And then a user could do `get_info('blas_opt').update(get_info('lapack_mkl'))` for instance (dummy example ;) )

File: numpy/distutils/system_info.py (hunk 7 above)
Comment (zerothi) [reply numpy/distutils/system_info.py]: Yes, it has been there for some time. But I dislike the above since one could put the same entries in the `libraries` key which would then allow the 2nd key.

In my opinion we should fix a `libraries` preference, and allow the older keywords (as exceptions). Once I have the error handler I will update.

File: numpy/distutils/system_info.py (hunk 6 above)
Comment (rgommers) [reply numpy/distutils/system_info.py]: I think splits are possible, they're different libraries (e.g. BLIS + Flame).

> ATLAS can provide either BLAS or BLAS+LAPACK(sub) or BLAS+LAPACK, with the middle option being only a subset of routines which are optimized.

I have never seen ATLAS used with numpy/scipy as anything other than "all of BLAS + LAPACK".

Comment (zerothi) [reply numpy/distutils/system_info.py]: @rgommers agreed, me neither. But the possibility *is* there.

**Comment by zerothi** [comment]
@eric-wieser, @rgommers I have marked conversations as resolved. Please reopen if needed!

File: numpy/distutils/system_info.py [hunk 8]
```diff
@@ -607,6 +611,35 @@ def set_info(self, **info):
...
+        """
+        found = list(map(lambda opt: self.cp.has_option(self.section, opt), options))
+        if sum(found) == 1:
+            return options[found.index(True)]
+        elif sum(found) == 0:
+            # nothing is found anyways
+            return options[0]
+
+        # Else we have more than 1 key found
+        raise AliasedOptionError(AliasedOptionError.__doc__.format(section=self.section, options='[{}]'.format(', '.join(options))))
```
Comment (mattip) [inline numpy/distutils/system_info.py]: 80 columns

File: numpy/distutils/system_info.py [hunk 9]
```diff
@@ -449,6 +449,10 @@ class NotFoundError(DistutilsError):
     """Some third-party program or library is not found."""
 
 
+class AliasedOptionError(DistutilsError):
+    """Aliases entries in config files should not be existing. In section '{section}' we found multiple of options {options} """
+
+
```
Comment (mattip) [inline numpy/distutils/system_info.py]: Needs reformat to 80 column width. Also "mulitple of options" might be better as "multiple appearances of option"

**Comment by mattip** [comment]
Could use some new tests in `numpy/distutils/tests/test_system_info.py`

**Comment by zerothi** [comment]
> Could use some new tests in `numpy/distutils/tests/test_system_info.py`

Added!

**Comment by zerothi** [comment]
@mattip fixed remaining issues!

**Comment by mattip** [comment]
It would be nice to document this but ...
* `site.cfg.example` is not linked into the html document tree at all.
* Documentation for `system_info` (which uses and describes `site.cfg`) is in the docstring for `numpy/distutils/system_info.py` but is not linked to the documentation, the closest we get is [`doc/DISTUTILS.rst.txt`](http://www.numpy.org/devdocs/reference/distutils_guide.html#numpy-distutils-system-info-module).

Could you take a look at both those files anyway (`site.cfg.example` and `system_info.py`) and update them for the preferred `libraries=` usage? Extra points if you figure out a nice way to link that into the documentation, but it should probably be a different issue/PR

File: site.cfg.example [hunk 10]
```diff
@@ -21,6 +21,8 @@
 #       with. Note that these should be just the names, not the filenames. For
 #       example, the file "libfoo.so" would become simply "foo".
 #           libraries = lapack,f77blas,cblas,atlas
+#       This setting is available for *all* sections, if not it should be
+#       considered a bug.
```
Comment (mattip) [inline site.cfg.example]: No need for the ", if not ..."

File: numpy/distutils/tests/test_system_info.py [hunk 11]
```diff
@@ -7,9 +7,10 @@
 from subprocess import Popen, PIPE
 from distutils.errors import DistutilsError
 
+from numpy.testing import assert_, assert_equal, assert_raises
 from numpy.distutils import ccompiler, customized_ccompiler
-from numpy.testing import assert_, assert_equal
 from numpy.distutils.system_info import system_info, ConfigParser
+from numpy.distutils.system_info import AliasedOptionError
```
Comment (mattip) [inline numpy/distutils/tests/test_system_info.py]: Collapse this line onto the previous one

File: numpy/distutils/system_info.py [hunk 12]
```diff
@@ -92,20 +92,20 @@
 search_static_first = 0
 
 [fftw]
-fftw_libs = rfftw, fftw
-fftw_opt_libs = rfftw_threaded, fftw_threaded
```
Comment (mattip) [inline numpy/distutils/system_info.py]: Why did you remove this line as part of this PR?

File: numpy/distutils/system_info.py [hunk 13]
```diff
@@ -92,20 +92,20 @@
...
-# if the above aren't found, look for {s,d}fftw_libs and {s,d}fftw_opt_libs
```
Comment (mattip) [inline numpy/distutils/system_info.py]: Why did you remove this comment as part of this PR?

File: numpy/distutils/system_info.py (hunk 12 above)
Comment (zerothi) [reply numpy/distutils/system_info.py]: `fftw_opt_libs` is never referenced anywhere in `system_info.py`, so I removed it due to its connection with `libraries` option

File: numpy/distutils/system_info.py (hunk 13 above)
Comment (zerothi) [reply numpy/distutils/system_info.py]: same as for previous comment

File: numpy/distutils/tests/test_system_info.py (hunk 11 above)
Comment (zerothi) [reply numpy/distutils/tests/test_system_info.py]: `<80` char limit

File: site.cfg.example (hunk 10 above)
Comment (zerothi) [reply site.cfg.example]: done

File: numpy/distutils/system_info.py (hunk 12 above)
Comment (mattip) [reply numpy/distutils/system_info.py]: interesting. Even when I go back to 575d373479 which added these lines, I do not see how it was ever parsed.

**Comment by mattip** [comment]
LGTM. Will give other reviewers a chance to take a look before merging

File: numpy/distutils/system_info.py (hunk 12 above)
Comment (zerothi) [reply numpy/distutils/system_info.py]: Agreed, I couldn't find any comments on it...

**Comment by mattip** [comment]
Thanks @zerothi
